### PR TITLE
draft: oas for testrun

### PR DESCRIPTION
**Pull request by thynquest**

### CODE REVIEW

File: packages/marble-api/openapis/marblecore-api.yaml [hunk 1]
```diff
@@ -3269,6 +3271,126 @@ paths:
...
+      security:
+        - bearerAuth: []
+      parameters:
+        - name: scenarioId
+          description: ID of the scenario that need to be fetched
+          in: query
+          required: true
+          schema:
+            type: string
+            format: uuid
```
Comment (balzdur) [inline packages/marble-api/openapis/marblecore-api.yaml:3289]: Since we do need to GET testruns by ID, I changed the signature from
```
GET /scenario-testrun/{scenarioId}
```
To
```
GET /scenario-testrun?scenarioId=... // use of a query param
```

Let me know if it's ok.

Another solution is to move everything link to a testrun under something like
```
GET /scenario/{scenarioId}/testrun
POST /scenario/{scenarioId}/testrun
...
```
But it doesn't make sens to fill the `{scenarioId}` on API endpoint using `{testRunId}`

File: packages/marble-api/openapis/marblecore-api.yaml [hunk 2]
```diff
@@ -3269,6 +3271,126 @@ paths:
...
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/TestRunDto'
+        '401':
+          $ref: '#/components/responses/401'
+        '403':
+          $ref: '#/components/responses/403'
+        '404':
+          $ref: '#/components/responses/404'
```
Comment (balzdur) [inline packages/marble-api/openapis/marblecore-api.yaml:3358]: Required for the testrun detail page.
The current endpoint is not supposed to return graph data (the APi for this is out of scope of this PR, existing work in [this openned backend PR](https://github.com/checkmarble/marble-backend/pull/746))

File: packages/marble-api/openapis/marblecore-api.yaml [hunk 3]
```diff
@@ -3269,6 +3271,126 @@ paths:
...
```
Comment (balzdur) [inline packages/marble-api/openapis/marblecore-api.yaml:3393]: This is the endpoint I'm totally not fan of (any proposition is welcomed).

The ultimate goal is to allow to "down" an ongoing testrun.

Cancelling ongoing testrun looks necessary when you try to create a new one.
Still, this can be handled by the backend usecase, and we can remove this from the API. It won't prevent the frontend to display a warning modal (more or less like the information modal displayed on draft creation)

File: packages/marble-api/openapis/marblecore-api.yaml [hunk 4]
```diff
@@ -5214,6 +5336,75 @@ components:
...
+          type: string
+          format: date-time
+        end_date:
+          type: string
+          format: date-time
+        creator_id:
+          type: string
+          format: uuid
+        status:
+          $ref: '#/components/schemas/TestRunStatusDto'
```
Comment (balzdur) [inline packages/marble-api/openapis/marblecore-api.yaml:5376]: This is the full list of what the frontend expect (according to the Figma).

Still, one thing is not included: an optionnal "log" info to informa user on the "why" an existing testrun transition to "down" status (ex: Manually cancelled, Cancelled due to live version update on the scenario, cancelled by another testrun....)

From the Figma, this part of the design (choice for now is to make this column out of scope for the first iteration on testrun):
image

File: packages/marble-api/openapis/marblecore-api.yaml [hunk 5]
```diff
@@ -5214,6 +5336,75 @@ components:
...
+          format: uuid
+        phantom_iteration_id:
+          type: string
+          format: uuid
+        start_date:
+          type: string
+          format: date-time
+        end_date:
+          type: string
+          format: date-time
```
Comment (balzdur) [inline packages/marble-api/openapis/marblecore-api.yaml:5400]: We can discuss on this DTO :
- `ref_iteration_id` seems unecessary with current business rules (= the only accepted version is the current live one) but we can still provide this in the payload and check this is the live_id on the backend (so in a future version, the API is ready to allow to pick a pair of version).
- `start_date` seems unecessary (unwanted)

**Review by balzdur — COMMENTED** [review]
I merged the work directly inside the `marblecore-api.yaml` file.

I make some changes on the DTO, lot of them require to be reviewed.

File: packages/marble-api/openapis/marblecore-api.yaml (hunk 1 above)
Comment (thynquest) [reply packages/marble-api/openapis/marblecore-api.yaml:3289]: the changed signature is ok for me. for the moment let's keep it simple, we will be able to change of needed. I will adapt the backend accordingly

File: packages/marble-api/openapis/marblecore-api.yaml (hunk 5 above)
Comment (Pascal-Delange) [reply packages/marble-api/openapis/marblecore-api.yaml:5400]: agreed on this - a test run should only be created on the currently live version (and cannot be created if none is live) - so we can remove this from the endpoint.
Similarly, I'd not allow to set a start date in the future - just use "now" (so no need to expose this either).
Finally, I'd go for just "scenario_iteration_id" (or if it's too ambiguous, "test_iteration_id" or such) instead of "phantom_iteration_id" : i find the concept of "phantom decisions" to be a convenient concept and naming to work with on the backend side, but IMO it's an implementation detail and I'd rather avoid "leaking" this naming into the rest of the product.

File: packages/marble-api/openapis/marblecore-api.yaml (hunk 3 above)
Comment (thynquest) [reply packages/marble-api/openapis/marblecore-api.yaml:3393]: let's decide about this later.. I think it is a product design issue; it is not necessary for the v0

File: packages/marble-api/openapis/marblecore-api.yaml (hunk 2 above)
Comment (thynquest) [reply packages/marble-api/openapis/marblecore-api.yaml:3358]: done .. I will integrate it

File: packages/marble-api/openapis/marblecore-api.yaml (hunk 4 above)
Comment (thynquest) [reply packages/marble-api/openapis/marblecore-api.yaml:5376]: works for me.

**Review by Pascal-Delange — APPROVED** [review]
